common/src/main/java/io/netty/util/concurrent/Future.java

```diff
@@ -306,6 +307,52 @@
     @Override
     boolean cancel(boolean mayInterruptIfRunning);
 
+    /**
+     * Create a new future that will complete with the result of this future mapped through the given mapper function.
+     * <p>
+     * If this future fails, then the returned future will fail as well, with the same exception.
+     * Cancellation of either future will cancel the other.
+     * If the mapper function throws, the returned future will fail, but this future will be unaffected.
+     *
+     * @param mapper The function that will convert the result of this future into the result of the returned future.
+     * @param <R> The result type of the mapper function, and of the returned future.
+     * @return A new future instance that will complete with the mapped result of this future.
+     */
+    default <R> Future<R> map(Function<V, R> mapper) {
+        DefaultPromise<R> promise = new DefaultPromise<>(executor());
+        addListener(PromiseOperator.map(mapper), promise);
+        promise.addListener(this, PromiseNotifier::propagateCancel);
+        return promise;
+    }
+
+    /**
+     * Create a new future that will complete with the result of this future flat-mapped through the given mapper
+     * function.
+     * <p>
+     * The "flat" in "flat-map" means the given mapper function produces a result that itself is a future-of-R,
+     * yet this method also returns a future-of-R, rather than a future-of-future-of-R.
+     * In other words, if the same mapper function was used with the {@link #map(Function)} method, you would get back
+     * a {@code Future<Future<R>>}.
+     * These nested futures are "flattened" into a {@code Future<R>} by this method.
+     * Note that the future returned by this method is not the same instance as the one the mapper function returns.
+     * The reason is that this method needs to return immediately, but the mapper function cannot be applied before this
+     * future has completed.
+     * <p>
+     * If this future fails, then the returned future will fail as well, with the same exception.
+     * Cancellation of either future will cancel the other.
+     * If the mapper function throws, the returned future will fail, but this future will be unaffected.
+     *
+     * @param mapper The function that will convert the result of this future into the result of the returned future.
+     * @param <R> The result type of the mapper function, and of the returned future.
+     * @return A new future instance that will complete with the mapped result of this future.
+     */
+    default <R> Future<R> flatMap(Function<V, Future<R>> mapper) {
+        DefaultPromise<R> promise = new DefaultPromise<>(executor());
+        addListener(PromiseOperator.flatMap(mapper), promise);
```

Comment:
I find going through the context route harder to follow than to just have a listener like:

```java
addListener(future -> {
    if (!promise.setUncancellable()) {
        return;
    }
    if (future.isSuccess()) {
        promise.setSuccess(mapper.apply(future.getNow()));
    } else {
        promise.setFailure(future.cause());
    }
});
```

IIUC the allocation profile does not change and we can always have static listener implementations to aid debugging as you are doing with the `PromiseOperator`